### PR TITLE
feat: Tests, better logging, naive responses, response caching

### PR DESCRIPTION
**Pull request by zoni527**
Introduces first Python test programs, that can be customized to send different kinds of correct and incorrect HTTP requests to the server (assumes port 8081 currently). Modified Server and Requests a bit, mostly updates to the logging, and some to organization, and expanding the interface for the Response forming to use. Responses are now sent to the client fd without any further validation, chunked sending doesn't exist yet either. Responses now also get cached in memory for faster access, only one read from file necessary. If cache fills up should perform cleanup, needs more testing though, haven't considered all edge cases.

### CODE REVIEW

File: srcs/Utils.cpp [hunk 1]
```diff
@@ -182,6 +188,39 @@ bool	uriTargetAboveRoot(std::string_view uri)
...
+
+	std::stringstream	buf;
+
+	buf << fileStream.rdbuf();
+	return buf.str();
+}
+
+std::string	getAbsPath(std::string const &fileName, std::string searchDir)
+{
+	static std::string const	currentDir = std::filesystem::current_path();
```
[the remark this answers is not in the document]
Comment (zoni527) [reply srcs/Utils.cpp]: It is

File: srcs/Server.cpp [hunk 2]
```diff
@@ -254,14 +264,34 @@ void	Server::handleClientData(size_t& i)
...
 			}
+			DEBUG_LOG("Keep alive status: " + std::to_string((*it).getKeepAlive()));
 			if (!(*it).getKeepAlive()) {
+				INFO_LOG("Closing fd " + std::to_string(_pfds[i].fd));
 				close(_pfds[i].fd);
+				INFO_LOG("Erasing fd " + std::to_string(_pfds[i].fd) + " from clients list");
+				_clients.erase(it);
 				if (_pfds.size() > (i + 1)) {
+					DEBUG_LOG("Overwriting index " + std::to_string(i) + " with " + std::to_string(_pfds.size() - 1));
+					INFO_LOG("Removing client fd " + std::to_string(_pfds[i].fd) + " from poll list");
```
[the remark this answers is not in the document]
Comment (zoni527) [reply srcs/Server.cpp]: Clarified message to talk about fds, not indices.

File: srcs/Response.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	//	if ok
+	//		get correct config
+	//		get correct content
+	//			from file or from memory?
+	//			What if the content is huge? Chunking time?
+
+	_target = req.getTarget();
+
+	if (std::filesystem::is_directory(_target)) {
+		if (_target.back() == '/')
```
[the remark this answers is not in the document]
Comment (zoni527) [reply srcs/Response.cpp]: We need to split of validation and error classification into its own step, this is currently already being validated by the Request forming, but we should consider what the best approach for it would be. Of course if there is a glaring error at the beginning it doesn't make much sense to keep parsing.

File: srcs/Utils.cpp [hunk 4]
```diff
@@ -182,6 +188,39 @@ bool	uriTargetAboveRoot(std::string_view uri)
...
+
+	if (searchDir.empty())
+		searchDir = currentDir;
+	if (searchDir.back() == '/')
```
[the remark this answers is not in the document]
Comment (zoni527) [reply srcs/Utils.cpp]: How the hell would currentDir ever be empty? Even if we were at root this wouldn't happen.

**Comment by ThiwankaS** [comment]
Will go ahead with this current implementation

**Review by ThiwankaS — APPROVED** [review]
We can go ahead with the current implementaion.

**Comment by zoni527** [comment]
> In the directory traversing code, we could have dircetly look insdie the root directory without passing the search directory as a parameter, but accounting furture code reusability will go ahead with the current implementation. The Python testing part will also be left for the future.

If the parameter is left empty the root directory where the program is being run from is automatically selected, or did you mean actual root /?
